src/navigation/perfect_memory_window.cc

```diff
@@ -0,0 +1,103 @@
+#include "navigation/perfect_memory_window.h"
+
+// Standard C++ includes
+#include <algorithm>
+
+namespace BoBRobotics {
+namespace Navigation {
+namespace PerfectMemoryWindow {
+
+//----------------------------------------------------------------------------
+// BoBRobotics::Navigation::PerfectMemoryWindow::Fixed
+//----------------------------------------------------------------------------
+Fixed::Fixed(size_t fwdLASize, size_t revLASize)
+:   m_FwdLASize(fwdLASize), m_RevLASize(revLASize),
+    m_MemoryPointer(std::numeric_limits<size_t>::max())
+{
+}
+//----------------------------------------------------------------------------
+std::pair<size_t, size_t> Fixed::getWindow() const
+{
+    // If this is our first update, return a full window
+    if(m_MemoryPointer == std::numeric_limits<size_t>::max()) {
+        return std::make_pair(0, std::numeric_limits<size_t>::max());
+    }
+    // Otherwise, return a window centred around memory memory pointer
+    // **NOTE** because size_t is unsigned, we need to be careful at the minimum
+    else {
+        return std::make_pair(
+            (m_MemoryPointer >= m_RevLASize) ? (m_MemoryPointer - m_RevLASize) : 0,
+            m_MemoryPointer + m_FwdLASize);
+    }
+}
+//----------------------------------------------------------------------------
+void Fixed::updateWindow(size_t bestSnapshot, float)
+{
+    m_MemoryPointer = bestSnapshot;
+}
+//----------------------------------------------------------------------------
+void Fixed::resetWindow()
+{
+    m_MemoryPointer = std::numeric_limits<size_t>::max();
+}
+
+//----------------------------------------------------------------------------
+// BoBRobotics::Navigation::PerfectMemoryWindow::DynamicBestMatchGradient
+//----------------------------------------------------------------------------
+DynamicBestMatchGradient::DynamicBestMatchGradient(size_t fwdLASize, size_t fwdLAIncreaseSize, size_t fwdLADecreaseSize, size_t minFwdLASize, size_t maxFwdLASize,
+                                                   size_t revLASize, size_t revLAIncreaseSize, size_t revLADecreaseSize, size_t minRevLASize, size_t maxRevLASize)
+:   m_MinFwdLASize(minFwdLASize), m_MinRevLASize(minRevLASize),
+    m_MaxFwdLASize(maxFwdLASize), m_MaxRevLASize(maxRevLASize),
+    m_FwdLAIncreaseSize(fwdLAIncreaseSize), m_RevLAIncreaseSize(revLAIncreaseSize),
+    m_FwdLADecreaseSize(fwdLADecreaseSize), m_RevLADecreaseSize(revLADecreaseSize),
+    m_FwdLASize(fwdLASize), m_RevLASize(revLASize),
+    m_LastLowestDifference(0.0f), m_MemoryPointer(std::numeric_limits<size_t>::max())
+{
+}
+//----------------------------------------------------------------------------
+std::pair<size_t, size_t> DynamicBestMatchGradient::getWindow() const
+{
+    // If this is our first update, return a full window
+    if(m_MemoryPointer == std::numeric_limits<size_t>::max()) {
+        return std::make_pair(0, std::numeric_limits<size_t>::max());
```

Comment:
FYI Nothing wrong with doing it this way, but you could also just use braced initialisation here for terseness: ``return {0, std::numeric_limits<size_t>::max()};``